geostore/s3_utils.py

```diff
@@ -52,3 +54,58 @@ def s3_url_reader(staging_url: str) -> GeostoreS3Response:
     staging_s3_client = get_s3_client_for_role(s3_role_arn)
     geostore_s3_client = get_s3_client_for_role(get_param(ParameterName.S3_USERS_ROLE_ARN))
     return s3_url_reader
+
+
+def check_if_s3_object_exists(bucket: str, key: str) -> bool:
+    s3_resource = boto3.resource("s3")
+
+    try:
+        s3_resource.Object(bucket, key).load()
+    except ClientError:
+        # 404 is most likely the response code here
+        # other response code is possible, but we return false rather than raise,
+        # thus allowing the next step to continue rather than stalling the entire process
+        return False
+
+    return True
+
+
+def get_s3_etag(s3_bucket: str, s3_object_key: str, logger: Logger) -> str:
+    s3_client = boto3.client("s3")
+
+    try:
+        s3_response = s3_client.head_object(Bucket=s3_bucket, Key=s3_object_key)
+        s3_etag = s3_response["ETag"].strip('"')
+        return s3_etag
+    except ClientError as error:
+        logger.error(
+            f"Unable to fetch eTag for {s3_object_key} in s3://{s3_bucket},"
+            f"due to {error}, even though object exists in bucket.",
+            extra={GIT_COMMIT: get_param(ParameterName.GIT_COMMIT)},
+        )
+        # rather than raise, we return an empty string, indicating that the etag is different
+        # thus allowing the next step to continue rather than stalling the entire process
+        return ""
+
+
+def calculate_s3_etag(
+    body: bytes,
+    # https://docs.aws.amazon.com/cli/latest/topic/s3-config.html#multipart-chunksize
+    chunk_size: int = 8 * 1024 * 1024,  # Default value
+) -> str:
+
+    md5s = []
+    data = [body[i : i + chunk_size] for i in range(0, len(body), chunk_size)]
```

Comment:
This is a good point. Is the concern here mainly around large files? Or is there any other reason to avoid doing so?
(i.e. if user submit files over 1gb, all of it would be written to memory, exceeding lambda limit?)

Reply:
Failing the job because of big files would definitely be an issue. Fortunately that's one we could work around (at least in the short term) by bumping the max memory and paying more for the Lambdas. Stream processing is also usually faster than bulk file processing, so there's also a performance concern.